### PR TITLE
fix: image disappear issue

### LINKED ISSUE #19546: WYSIWYG: Image disappears after going out of editing mode
- Azure Data Studio Version: 1.36.2
- OS Version: Windows

Steps to Reproduce:

1. Paste an image
2. Go out of editing mode by clicking anywhere outside the cell

repro content being sent to the team

Does this issue occur when all extensions are disabled?: Yes/No

### PR DESCRIPTION
**Pull request by MaddyDev**
This PR fixes #19546 

Issue: When user pastes an image in wysiwyg mode and goes out of the edit mode, the image disappears.
Why: When user pastes an image into a notebook that already has html tags in it, the image is added as html inside span tags. And when we parse for attachments, we're replacing the valid base64 value of the image with attachment:imageName inside the image tags src attribute, which renders the image broken.

Fix: Check if the image is added as an html image element before adding it to the list of attachments of the cell. ( By checking if the valid base64 value being added to the attachments is in an image tag)
Also tried to add it as an actual attachment and replacing the entire image tag with markdown attachment syntax. But since that's embedded in-between other html tags, it doesn't seem to work. So refrained from adding it as an attachment and just skipping it.

### CODE REVIEW

File: src/sql/workbench/services/notebook/browser/models/cell.ts [hunk 1]
```diff
@@ -349,8 +349,9 @@ export class CellModel extends Disposable implements ICellModel {
 	private attachImageFromSource(newSource: string | string[]): string | string[] {
 		if (!Array.isArray(newSource) && this.isValidBase64OctetStream(newSource)) {
 			let results;
-			while ((results = validBase64OctetStreamRegex.exec(newSource)) !== null) {
-				let imageName = this.addAttachment(results[1], results[0], 'image.png');
+			// only replace the base64 value if it's from markdown [](base64value) not html tags <img src="base64value"
+			while ((results = validBase64OctetStreamRegex.exec(newSource)) !== null && newSource.substring(newSource.indexOf(results[0]) - 9, newSource.indexOf(results[0]) - 2) !== 'img src') {
+				let imageName = this.addAttachment(results[1], results[0], 'image0.png');
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/notebook/browser/models/cell.ts:354]: Why did you change the image name?

File: src/sql/workbench/services/notebook/browser/models/cell.ts [hunk 2]
```diff
@@ -349,8 +349,9 @@ export class CellModel extends Disposable implements ICellModel {
 	private attachImageFromSource(newSource: string | string[]): string | string[] {
 		if (!Array.isArray(newSource) && this.isValidBase64OctetStream(newSource)) {
 			let results;
-			while ((results = validBase64OctetStreamRegex.exec(newSource)) !== null) {
-				let imageName = this.addAttachment(results[1], results[0], 'image.png');
+			// only replace the base64 value if it's from markdown [](base64value) not html tags <img src="base64value"
+			while ((results = validBase64OctetStreamRegex.exec(newSource)) !== null && newSource.substring(newSource.indexOf(results[0]) - 9, newSource.indexOf(results[0]) - 2) !== 'img src') {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/notebook/browser/models/cell.ts:353]: Why not update the regex to only look at [](base64Value) values instead? That'd be cleaner and remove the chance of some other format breaking this.

Comment (chlafreniere) [inline src/sql/workbench/services/notebook/browser/models/cell.ts:353]: Instead of checking for this logic here, can we consolidate logic in notebookMarkdown.ts? https://github.com/microsoft/azuredatastudio/blob/main/src/sql/workbench/contrib/notebook/browser/outputs/notebookMarkdown.ts#L74

Haven't looked into this myself, just took a quick look here. Ideally, the cell model knows about attachments, but doesn't have to actually do any html parsing.

[the remark this answers is not in the document]
Comment (MaddyDev) [reply src/sql/workbench/services/notebook/browser/models/cell.ts:353]: good point, let me add it to the regex, it technically should not have src="base64Value" so just checking that in the regex should work.

Comment (MaddyDev) [reply src/sql/workbench/services/notebook/browser/models/cell.ts:353]: I tried that but it doesn't hit that flow at all when an image is pasted. Going to debug and check where it's building the html part.

File: src/sql/workbench/services/notebook/browser/models/cell.ts [hunk 3]
```diff
@@ -349,7 +349,10 @@ export class CellModel extends Disposable implements ICellModel {
 	private attachImageFromSource(newSource: string | string[]): string | string[] {
 		if (!Array.isArray(newSource) && this.isValidBase64OctetStream(newSource)) {
 			let results;
-			while ((results = validBase64OctetStreamRegex.exec(newSource)) !== null) {
+			// only replace the base64 value if it's from markdown [](base64value) not html tags <img src="base64value">
+			let validImageTag = /<img\s+[^>]*src="([^"]*)"[^>]*>/;
```
[the remark this answers is not in the document]
Comment (MaddyDev) [reply src/sql/workbench/services/notebook/browser/models/cell.ts:353]: It would be a malformed tag if there is a > inside the elements attributes.

[the remark this answers is not in the document]
Comment (Charles-Gagnon) [reply src/sql/workbench/services/notebook/browser/models/cell.ts:353]: Not necessarily - e.g. [https://github.com/Microsoft/azuredatastudio/issues/new?title=Test<>Test](https://github.com/Microsoft/azuredatastudio/issues/new?title=Test<>Test) is a valid URL, so you could have an image with something like that as a URL.

[the remark this answers is not in the document]
Comment (Charles-Gagnon) [reply src/sql/workbench/services/notebook/browser/models/cell.ts:353]: The link was just an example URL since I don't have an image available that has a <> in the URL - but for example an image with a URL such as `https://test.com/myimage.png?somedata=Test<>Test` would be possible for someone to enter. 

So if I paste this into the text cell : `Test" >` is that URL going to be escaped before it gets to this logic?

Granted this is getting into edge case territory, but we need to make sure we're at least aware of these issues and trying to handle them if we can (and if not then making it clear that that's intended behavior for now)

[the remark this answers is not in the document]
Comment (Charles-Gagnon) [reply src/sql/workbench/services/notebook/browser/models/cell.ts:353]: Here's an actual URL you can use for testing : [https://raw.githubusercontent.com/microsoft/azuredatastudio/main/resources/win32/code_150x150.png?somedata=<>](https://raw.githubusercontent.com/microsoft/azuredatastudio/main/resources/win32/code_150x150.png?somedata=<>)

[the remark this answers is not in the document]
Comment (MaddyDev) [reply src/sql/workbench/services/notebook/browser/models/cell.ts:353]: For the scenario am trying to fix, it's very specific to copy paste images and the src value will always be a valid base64 octet value.

[the remark this answers is not in the document]
Comment (Charles-Gagnon) [reply src/sql/workbench/services/notebook/browser/models/cell.ts:353]: Yeah, good point. Even though the regex would get the incorrect value if it has a > in it that shouldn't matter. It would be a good idea to call that out specifically in a comment though so later on we don't try to re-use this for something that does actually care about handling absolute URLs.

File: src/sql/workbench/services/notebook/browser/models/cell.ts [hunk 4]
```diff
@@ -349,7 +349,11 @@ export class CellModel extends Disposable implements ICellModel {
...
+			let imageResults;
+			// Note: It would not add image attachments from makrdown below the html tags.
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/notebook/browser/models/cell.ts:355]: `Note: Currently this will not process any markdown image attachments that are below an HTML img element. This is acceptable for now given the low risk of this happening and an easy workaround being to just changing the img element to a markdown embedded image instead`

**Review by Charles-Gagnon — APPROVED** [review]
Some comment update suggestions but the logic should be fine enough given that this is only really meant to fix a very specific edge case scenario.
